### PR TITLE
Return non-zero status when tests fail

### LINKED ISSUE #9: Report a non-zero exit code when tests fail
When tests fail, the process should exit with a non-zero exit code. This is especially important for people using this in CI.

Current behavior

```
$ clj -Atest

Running tests in #{"test"}

Testing foo

FAIL in (foo-test) (foo.clj:5)
expected: false
  actual: false

Ran 1 tests containing 1 assertions.
1 failures, 0 errors.

$ echo $?
0
```

Expected

```
$ echo $?
1
```

### PR DESCRIPTION
**Pull request by Olical**
Fix #9

### CODE REVIEW

**Comment by puredanger** [comment]
What if it just returned the `fail` count as the exit code?

**Comment by Olical** [comment]
Hmm, kinda neat since you get more information as well as an elegant piece of code as a solution. Buuuut, isn't that abusing the exit status and overloading it with more information than it was intended for?

After reading this thread, I feel like it's kind of abusing it for the wrong reasons even more https://unix.stackexchange.com/questions/418784/what-is-the-min-and-max-values-of-exit-codes-in-linux

In my opinion, the only thing we're encoding is "it will be non-zero when something went wrong". By convention, this usually means 1. Theoretically we could have 2000 tests fail though, right? Which seems like it could go horribly wrong in some strange edge cases.

**Comment by elzibubble** [comment]
If you return the failed count as exit code, when 256 tests fail it'll wrap to 0.

**Comment by lwhorton** [comment]
The summary is currently spitting out
```
{:test 3, :pass 2, :fail 0, :error 1, :type :summary}
```
What's the distinction between a failing test and an error? Would we want a 1 exit code on `(not (zero? (+ fail error)))`?

**Comment by puredanger** [comment]
Fail = assertion failed
Error = threw exception

**Comment by aviflax** [comment]
Any chance this might get merged soon?

**Comment by Olical** [comment]
I realised a while back that I'm killing the process which would be a pretty nasty breaking change for existing users that may run this from a REPL or something. So maybe I need to make some changes around that part.

I may have to check `:fail` as well as `:error` counts too, maybe just one isn't the right way. Like if `:fail` is 0 but there was an error it'll still think everything is okay and return 0.

**Comment by aviflax** [comment]
Ah yeah that makes sense. Thanks!

**Comment by Olical** [comment]
I've just rebased the branch ontop of master. If we want to check `fail` and `error` either myself or someone else can apply this patch:

```diff
diff --git a/src/cognitect/test_runner.clj b/src/cognitect/test_runner.clj
index b6e9bbf..96eaaac 100644
--- a/src/cognitect/test_runner.clj
+++ b/src/cognitect/test_runner.clj
@@ -114,7 +114,7 @@
       (if (-> args :options :test-help)
         (help args)
         (try
-          (let [{:keys [fail]} (test (:options args))]
-            (exit (if (zero? fail) 0 1)))
+          (let [{:keys [fail error]} (test (:options args))]
+            (exit (if (zero? (+ fail error)) 0 1)))
           (finally
             (shutdown-agents)))))))
```

**Comment by aviflax** [comment]
That looks good! I’m just a little confused about the patch? Why not just apply this change to your branch in your fork, so it’d be included in the PR?

**Comment by Olical** [comment]
Oh I can do, but I don't know if it's the right thing to do. So if it's
fine in the current form someone can just hit merge. If that change is
require it's easy to apply.

**Comment by aviflax** [comment]
Ah cool, yeah, that makes sense. Thanks for explaining!

BTW — this is no big deal at all, just thought you might be interested — I’ve been using your branch in my project, and my latest CI build failed because the git SHA I was using has now disappeared, since you rebased the branch. It’s interesting to consider that now that people might be using any commit in any repo in a Clojure project, perhaps we should think twice before rebasing or rewriting any pushed commits!

**Comment by Olical** [comment]
Oh no! That's really bad, I'm sorry about that! I don't usually rebase I just know a lot of projects prefer it, I usually just merge back into my branch if I need things from master.

I _assumed_ the commit would hang around but not be referenced by anything, apparently I was wrong :sob: force pushing catches me out once again!

Note: I have updated my two posts that mentioned that old sha, they now point at the new one. Having commits vanish could definitely become annoying though. Although I shouldn't have just force pushed like I did, that was silly of me. I can't find any other mention of the old sha on Google.

**Comment by aviflax** [comment]
No worries! This is a new situation that we’re in the early stages of learning about and adapting to. I for one remain excited by the idea of referencing and including dependencies as _code_ rather than artifacts, using git shas rather than version numbers (which IMHO should be reserved for marketing and reflection, not for identifying revisions).

**Comment by levand** [comment]
Fix in another PR. Thanks!
